core/src/main/java/org/springframework/security/authorization/method/PostAuthorizeReactiveAuthorizationManager.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2002-2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.authorization.method;
+
+import org.aopalliance.intercept.MethodInvocation;
+import reactor.core.publisher.Mono;
+
+import org.springframework.security.access.expression.method.MethodSecurityExpressionHandler;
+import org.springframework.security.access.prepost.PostAuthorize;
+import org.springframework.security.authorization.AuthorizationDecision;
+import org.springframework.security.authorization.ReactiveAuthorizationManager;
+import org.springframework.security.core.Authentication;
+
+/**
+ * A {@link ReactiveAuthorizationManager} which can determine if an {@link Authentication}
+ * has access to the returned object from the {@link MethodInvocation} by evaluating an
+ * expression from the {@link PostAuthorize} annotation.
+ *
+ * @author Evgeniy Cheban
+ */
+public final class PostAuthorizeReactiveAuthorizationManager
+		implements ReactiveAuthorizationManager<MethodInvocationResult> {
+
+	private final PostAuthorizeExpressionAttributeRegistry registry = new PostAuthorizeExpressionAttributeRegistry();
+
+	/**
+	 * Sets the {@link MethodSecurityExpressionHandler}.
+	 * @param expressionHandler the {@link MethodSecurityExpressionHandler} to use
+	 */
+	public void setExpressionHandler(MethodSecurityExpressionHandler expressionHandler) {
```

Comment:
Can you elaborate on why immutability is wanted in this instance? It seems common in Spring Security to use setters for fields that have a default value and the constructor for fields that require a value from the application. Since there is a default value for the expression handler, a setter is more in line with the convention.

Reply:
It's a little misleading to have a final variable that is mutable (that means it really isn't mutable). We should prefer immutability if possible as it is thread safe. By using this approach we can also make PostAuthorizeExpressionAttributeRegistry immutable.

Reply:
This seems to touch on a broader conversation about how we treat optional values in Spring Security. If I'm following your argument correctly, every Spring bean that has configuration setters is by the same (correct) definition not thread-safe. So I guess I'm not quite seeing what makes this component special enough that we should break the prevailing convention that seems to embrace the thread-safety tradeoff.

Or, IOW, I'm okay with taking steps to start favoring constructor injection for optional as well as required members, but it feels like it should be taken up separately unless the problem presented by this class is uniquely problematic compared to all the other places in Spring Security where we do the same thing.

Reply:
The difference here is that I find it misleading that `PostAuthorizeExpressionAttributeRegistry` is final but can be modified by `setExpressionHandler`. Typically we try and ensure that final variables are immutable as well.

Another concern is that if we want to refactor it to later perhaps supporting an `ExpressionAttributeRegistry` then having the setter causes confusion since setting the registry and the MethodSecurityHandler would be mutually exclusive.